b2/b2http.py

```diff
@@ -0,0 +1,178 @@
+######################################################################
+#
+# File: b2/b2http.py
+#
+# Copyright 2016 Backblaze Inc. All Rights Reserved.
+#
+# License https://www.backblaze.com/using_b2_code.html
+#
+######################################################################
+
+from __future__ import print_function
+
+import json
+import requests
+import six
+import socket
+
+
+class WebApiError(Exception):
+    """
+    Holds information extracted from a call to a B2 API.
+
+    When an error happens, B2 returns an error JSON that looks like this:
+
+        {
+            "status" : 400,
+            "code" : "invalid_bucket_name",
+            "message" : "bucket name is too long"
+        }
+    """
+
+    def __init__(self, status, code, message):
+        super(WebApiError, self).__init__(status, code, message)
+        self.status = status
+        self.code = code
+        self.message = message
+
+    def __repr__(self):
+        return "WebApiError(%d, '%s', '%s')" % (self.status, self.code, self.message)
+
+    def __eq__(self, other):
+        return repr(self) == repr(other)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+def _print_exception(e, indent=''):
+    """
+    Used for debugging to print out nested exception structures.
+    """
+    print(indent + 'EXCEPTION', repr(e))
+    print(indent + 'CLASS', type(e))
+    for (i, a) in enumerate(e.args):
+        print(indent + 'ARG %d: %s' % (i, repr(a)))
+        if isinstance(a, Exception):
+            _print_exception(a, indent + '        ')
+
+
+def _translate_errors(fcn):
+    """
+    Calls the given function, turning any exception raised into a WebApiError.
+    """
+    try:
+        response = fcn()
+        if response.status_code not in [200, 206]:
+            # Decode the error object returned by the service
+            error = json.loads(response.content.decode('utf-8'))
+            raise WebApiError(int(error['status']), error['code'], error['message'])
+        return response
+
+    except WebApiError:
+        raise  # pass through exceptions from just above
+
+    except requests.ConnectionError as e0:
+        e1 = e0.args[0]
+        if isinstance(e1, requests.packages.urllib3.exceptions.MaxRetryError):
+            msg = e1.args[0]
+            if 'nodename nor servname provided, or not known' in msg:
+                # Unknown host, or DNS failing.  In the context of calling
+                # B2, this means that something is down between here and
+                # Backblaze, so we treat it like 503 Service Unavailable.
+                raise WebApiError(503, 'unknown_host', 'unable to locate host')
```

Comment:
This is misleading: 503 means good connection to load balancer/front server,  but backend server error/all backends broken or busy